src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1735,520 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+
+            size_t truncate_after = 0;
+            if (num_segments == 1) {
+                // There are two cases here.
+                // 1. The segment has at least target_size bytes after
+                // base_file_pos,
+                // 2. The segment is too small. In this case we need to clamp
+                // the result.
+                truncate_after = first_segment_file_pos + target.target_size;
+                truncate_after = std::clamp(
+                  truncate_after,
+                  first_segment_file_pos,
+                  it->get()->file_size());
+            } else {
+                // In this case we need to find the truncation point
+                // always starting from the beginning of the segment.
+                //
+                // prev is guaranteed to be smaller than target_size
+                // because we reached this branch.
+                auto prev = current_size - it->get()->file_size();
+                auto delta = target.target_size - prev;
+                truncate_after = delta;
+            }
+            auto last_index_entry = it->get()->index().find_above_size_bytes(
+              truncate_after);
+            if (
+              last_index_entry.has_value()
+              && model::prev_offset(last_index_entry->offset) > first) {
+                vlog(
+                  stlog.debug,
+                  "Setting offset range to {} - {}, {} bytes of the last "
+                  "segment are included",
+                  first,
+                  last_index_entry->offset,
+                  last_index_entry->filepos);
+
+                last_included_offset = model::prev_offset(
+                  last_index_entry->offset);
+
+                // We're including only part of the file so the total size
+                // of the range has to be adjusted
+                current_size
+                  -= (it->get()->file_size() - last_index_entry->filepos);
+            } else {
+                // There is no index entry that we can use to find the size
+                // of the offset range in this case
+                vlog(
+                  stlog.debug,
+                  "Setting offset range to {} - {} (end of segment), "
+                  "truncation point: {}, base_file_pos: {}",
+                  first,
+                  it->get()->offsets().committed_offset,
+                  truncate_after,
+                  first_segment_file_pos);
+                last_included_offset = it->get()->offsets().committed_offset;
+            }
+            break;
+        } else if (current_size > target.min_size) {
+            vlog(
+              stlog.debug,
+              "Setting offset range to {} - {}",
+              first,
+              it->get()->offsets().committed_offset);
+            // We can include full segment to the list of segments
+            last_included_offset = it->get()->offsets().committed_offset;
+            continue;
```

Comment:
Why is this gated on `current_size > target.min_size`? Doesn't the `current_size < target.min_size` check below kind of encompass that? Or put another way, if the first segments we iterated through didn't meet the `min_size` requirement, why is it important to avoid setting this?

Reply:
Got it, thanks for the explanation. IMO it makes it more confusing leaving it in, especially because there are so many branches in this code already